jupyter_console/interactiveshell.py

```diff
@@ -597,3 +644,23 @@ def init_history(self):
         self.history_manager = ZMQHistoryManager(client=self.client)
         self.configurables.append(self.history_manager)
 
+    def get_source_and_reset(self):
+        """
+        Get the source code entered so far, and reset source input line(s).
+        """
+        source, self._source_lines_buffered = self._source_lines_buffered, []
+        return "\n".join(source)
+
+    def handle_source(self, line):
+        """
+        Save the line with previous source lines, and see if more
+        input is needed. Returns more? and indent_string.
+        """
+        self._source_lines_buffered.append(line)
+        ## Ask client if line is complete; get indent for next line:
+        if self.use_kernel_is_complete:
+            msg_id = self.client.is_complete("\n".join(self._source_lines_buffered))
+            return self.handle_is_complete_reply(msg_id, timeout=0.05)
+        else:
+            more = (line != "")
```

Comment:
In case there's one day some auto-indenting interfering with this, I think we should test `line.strip()`, which will be the empty string if line contains only whitespace.